### ISSUE TITLE
wso2/gmail package : Access token fails to generate by given refresh token

### ISSUE THREAD

**Issue by MadhukaHarith92** [ballerina-platform/ballerina-lang#10668]
**Description:**
I am trying to send an email notification. It works fine when I my access token is not expired but fails to generate a new access token by the given refresh token which I obtained from oauthplayground (https://developers.google.com/oauthplayground). 

Then I implemented a function to generate a new access token. But it gives an error :"{"error":"unauthorized_client", "error_description":"Unauthorized"} " . 

**Steps to reproduce:**
function sendMail(string senderEmail, string receiverEmail, string subject, string messageBody) returns boolean {
    endpoint gmail:Client gmailClient {
        clientConfig:{
            auth:{
                accessToken:config:getAsString("gmailApiConfiguration.accessToken"),
                refreshToken:config:getAsString("gmailApiConfiguration.refreshToken"),
                clientId:config:getAsString("gmailApiConfiguration.clientId"),
                clientSecret:config:getAsString("gmailApiConfiguration.clientSecret")
            }   
        }
    };

    //Create html message
    gmail:MessageRequest messageRequest;
    messageRequest.recipient = receiverEmail;
    messageRequest.sender = senderEmail;
    messageRequest.subject = subject;
    messageRequest.messageBody = messageBody;
    messageRequest.contentType = gmail:TEXT_HTML;

    //Send mail
    var sendMessageResponse = gmailClient->sendMessage(senderEmail, untaint messageRequest);
    string messageId;
    string threadId;
    match sendMessageResponse {
        (string, string) sendStatus => {
            (messageId, threadId) = sendStatus;
            log:printInfo("Sent email to " + receiverEmail + " with message Id: " + messageId + " and thread Id:"
                    + threadId);
            return true;
        }
        gmail:GmailError e => {
            log:printInfo(e.message);
            return false;
        }
    }
}

public function refreshAccessToken() returns string {
    endpoint http:Client refreshTokenHTTPEP {
        url: "https://www.googleapis.com/oauth2/v4"
    };
    string accessToken;
    string refreshToken = config:getAsString("gmailApiConfiguration.refreshToken");
    string clientId = config:getAsString("gmailApiConfiguration.clientId");
    string clientSecret = config:getAsString("gmailApiConfiguration.clientSecret");

    http:Request httpRequest = new;
    //http:Response httpResponse = new;

    httpRequest.setHeader("Content-Type", "application/x-www-form-urlencoded");

    json jsonMsgrequest = { grant_type: "refresh_token", client_id: clientId, client_secret: clientSecret, refresh_token: refreshToken };
    httpRequest.setJsonPayload(jsonMsgrequest);
    //http:HttpConnectorError httpConnectorError;

    var httpPostReply = refreshTokenHTTPEP->post("/token", httpRequest);
    match  httpPostReply {
        http:Response httpResponse => {
            var jsonPayload = httpResponse.getJsonPayload();
            match jsonPayload {
                json response => {
                    try {
                        accessToken = response.access_token.toString();
                        log:printInfo("Response:");
                        log:printInfo(response.toString());
                    } catch (error err) {
                        log:printInfo("New access token error 1:");
                        log:printError(response.error.toString());
                    }
                }
                error e => {
                    log:printInfo("New access token error 2:");
                    log:printError(e.message);
                }
            }
        }
        error e => {
            log:printInfo("New access token error 3:");
            log:printError("Error in generating access token : " + e.message);
        }
    }

    return accessToken;
} 
**Affected Versions:**

ballerina : 0.982.0

**OS, DB, other environment details and versions:**

wso2/gmail : 0.9.7

**Related Issues (optional):**
https://github.com/ballerina-platform/ballerina-lang/issues/10641

**Suggested Labels (optional):**
wso2/gmail

**Suggested Assignees (optional):**

**Comment by MadhukaHarith92** [comment]
@Anoukh : The access token was successfully generated when we use our own oauth credentials (from https://console.developers.google.com) when obtaining refresh token from oauthplayground (settings in https://developers.google.com/oauthplayground/). However that step was not mentioned in https://central.ballerina.io/wso2/gmail . Not completely sure if that was the root cause though.

**Comment by Anoukh** [comment]
@MadhukaHarith92 When the access token expires, does a new token get generated using the refresh token?
Also, the method refreshAccessToken() is not required, right?

**Comment by MadhukaHarith92** [comment]
@Anoukh  : Yes I used an expired access token and a new one was generated using the refresh token. 
Yes refreshAccessToken() method was not required.
